kafka/integration_test.go

```diff
@@ -0,0 +1,529 @@
+	if err != nil {
+		t.Fatalf("Failed to create producer: %s\n", err)
+	}
+
+	t.Logf("Created Producer %v\n", p)
+
+	var doneChan chan bool
+
+	for _, testmsg := range testmsgs {
+
+		doneChan = make(chan bool)
+
+		// delivery handler
+		go func() {
+			e := <-p.Events()
+			verifyEvent(e, testmsg, t)
+			close(doneChan)
+		}()
+
+		m := &Message{
+			TopicPartition: TopicPartition{Topic: &testconf.Topic, Partition: testmsg.Partition},
+			Value:          testmsg.Value,
+			Key:            testmsg.Key}
+		p.ProduceChannel() <- m
+
+		// wait for delivery report goroutine to finish
+		_ = <-doneChan
+
+	}
+
+	p.Close()
+
+}
+
+// verify event info
+func verifyEvent(e Event, testmsg testmsgType, t *testing.T) {
+	switch e.(type) {
+	case *Message:
+		m := e.(*Message)
+		if m.TopicPartition.Error != nil {
+			err := m.TopicPartition.Error
+			if testmsg.ExpectedError != "" && err.Error() == testmsg.ExpectedError {
+				fmt.Fprintf(os.Stderr, "OK: Expected error: %s\n", testmsg.ExpectedError)
+			} else {
+				t.Errorf("Encountered unexepected error: %s\n", err)
+			}
+			return
+
+		}
+		// check partition
+		if !((testmsg.Partition == PartitionAny && m.TopicPartition.Partition >= 0) ||
+			(testmsg.Partition >= 0 && m.TopicPartition.Partition == testmsg.Partition)) {
+			t.Errorf("Partition check failed for test message %v. Got back {[%d] [%s] [%s]}\n",
+				testmsg, m.TopicPartition.Partition, m.Key, m.Value)
+			return
+
+		}
+
+		// check Key
+		if testmsg.Key == nil {
+			if !(m.Key == nil || len(m.Key) == 0) {
+				t.Errorf("Key verification failed. Expected %s, got %s\n", testmsg.Key, m.Key)
+				return
+			}
+		} else {
+			if string(m.Key) != string(testmsg.Key) {
+				t.Errorf("Key verification failed. Expected %s, got %s\n", testmsg.Key, m.Key)
+				return
+			}
+		}
+
+		// check Value
+		if testmsg.Value == nil {
+			if !(m.Value == nil || len(m.Value) == 0) {
+				t.Errorf("Value verification failed. Expected %s, got %s\n", testmsg.Value, m.Value)
+				return
+			}
+		} else {
+			if string(m.Value) != string(testmsg.Value) {
+				t.Errorf("Value verification failed. Expected %s, got %s\n", testmsg.Value, m.Value)
+				return
+			}
+		}
+
+		fmt.Printf("%% Message delivered to topic %s [%d] at offset %v: key=(%s), value=(%s)\n",
+			*m.TopicPartition.Topic, m.TopicPartition.Partition, m.TopicPartition.Offset, m.Key, m.Value)
+
+	default:
+		fmt.Printf("Ignored event: %s\n", e)
+	}
+}
+
+// TestConsumerFuncBasedAPI test function-based API, requires broker
+func TestConsumerFuncBasedAPI(t *testing.T) {
+	runConsumerFuncBasedAPI(t, "CommitMessage")
+	runConsumerFuncBasedAPI(t, "CommitOffsets")
+	runProducerChannelBasedAPI(t)
+	runConsumerFuncBasedAPI(t, "Commit")
+}
+
+func runConsumerFuncBasedAPI(t *testing.T, commitMode string) {
+	if !testconfRead() {
+		t.Skipf("Missing testconf.json")
+	}
+	c, err := NewConsumer(&ConfigMap{
+		"bootstrap.servers":    testconf.Brokers,
+		"group.id":             groupId,
```

Comment:
Since these tests will be run on someones broker we probably want the groupid to be configurable through testconf.json too